### DISCUSSION TITLE
How to fix quarto metadata being escaped like latex when compiling to typst?

### DISCUSSION THREAD

**Discussion by charging1948** [quarto-dev/quarto-cli#8983] [Q&A] [answered]
### Description

- Quarto 1.4.550
- Linux NixOS 23.11

### Discovery
I am trying to write a custom template using typst in quarto. The current version (at the time of writing) of this template can be found here:
[Template](https://github.com/Charging1948/quarto-typst-dhbw/tree/efe8f51b696436e01aeab6427a3c174624bea947)

While iterating on the template i have noticed that significant whitespace in certain strings and the @-symbol are being escaped.
For example:
'Prof. Dr.' -> 'Prof.~Dr.'
or
'test@' -> 'test\\@'

So far i have experienced this behavior when the relevant strings are inside the qmd-metadata at the top of file. Like this:
```yml
keep-typ: true
format:
  dhbw-typst: default
guide:
  name: "Prof. Dr. Testperson"
authors:
  - name: "Testperson"
    email: "test.test@test.me"
```

I think i have boiled down the problem enough to exclude typst as the culprit. Using `keep-typ: true` i have taken a look at the intermediate typst file and at this point the relevant portion looks like this:

```typ
#show: doc => project(
  authors: (
    ( name: "Testperson",
      email: "test.test\@test.me" ),
    ),
  guide: (
    name: "Prof.~Dr.~Testperson",
    email: "",
    department: "",
    affiliation: "",
),
  doc,
)
```

Is there any option to prevent this preprocessing? I am guessing this preprocessing is important for latex, as otherwise certain symbols would just vanish.
I cant remember having these problems when using the default typst template, though.

Relevant portions in final pdf look like this:

![image](https://github.com/quarto-dev/quarto-cli/assets/119578282/a905635b-b28a-4047-89ff-08e6899e0902)

Thanks in advance!

**Comment by mcanouil** [comment, marked as answer]
See <https://github.com/mcanouil/quarto-invoice/blob/main/_extensions/invoice/typst-template.typ#L59>.

```typst
object.replace("\\", "")
```

**Reply by charging1948** [reply]
Thank you very much for the quick response!

**Comment by cderv** [comment]
I was curious about Why this happens, and what is really the difference between default format, and your format. 

If I do use `format: typst` with some the name and email in your example 
````markdown
---
title: "Quarto Playground"
format: typst
keep-typ: true
authors:
  - name: "Prof. Dr. Testperson"
    email: "test.test@test.me"
---

Test
````

I got this in the `.typ`
````typst
#show: doc => article(
  title: [Quarto Playground],
  authors: (
    ( name: [Prof.~Dr.~Testperson],
      affiliation: [],
      email: [test.test\@test.me] ),
    ),
  toc_title: [Table of contents],
  toc_depth: 3,
  cols: 1,
  doc,
)
````

So I see the `~` and `\@`  too, but the pdf looks ok
![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/0832cbb1-28e7-4462-a7b7-64c5413b734e)

I believe this is because the `email` and `name` are not as string (using `".."` in this case) but are in `[..]` which is block content (https://typst.app/docs/reference/scripting/#blocks) 

This is from the template 
https://github.com/quarto-dev/quarto-cli/blob/248a3a4b1cd8985eb062cd53e53b23326c13eb2d/src/resources/formats/typst/pandoc/quarto/typst-show.typ#L5-L13

So it seems the escaping with `\@` is necessary in that case for `@` not to be taken as a reference. And `~` seems the way the space in interpreted by Pandoc in names (for unbreakable maybe ?)

Here is how Pandoc parse those in markdown to convert to typst

````powershell
> quarto pandoc -t typst
test.test@test.me
^Z
test.test\@test.me

> quarto pandoc -t typst
Prof. Dr. Testperson
^Z
Prof.~Dr.~Testperson

>  quarto pandoc -t native
Prof. Dr. Testperson
^Z
[ Para [ Str "Prof.\160Dr.\160Testperson" ] ]
````

Any YAML field will be interpreted by Pandoc as Markdown to be parsed. So this is why this gets like this. 

To prevent parsing, raw inline can be passed. For example, try this in your document
````yaml
---
format:
  dhbw-typst: default
guide:
  name: "`Prof. Dr. Testperson`{=typst}"
authors:
  - name: "Testperson"
    email: "`test.test@test.me`{=typst}"
keep-typ: true
---
````

The output should not have the `~` or the `\@`, and the value will be inserted in your template within the `".."` you have. 

So this behavior is a mix of 
* how Pandoc consider metadata 
	Doc is at [yaml_metadata_block](https://pandoc.org/MANUAL.html#extension-yaml_metadata_block) 
	> but all string scalars will be interpreted as Markdown")
* how the `.typ` template is using the Pandoc's variable 
	[interpolated variables](https://pandoc.org/MANUAL.html#interpolated-variables)
	> If the value of the variable is a simple value, it will be rendered verbatim.
* how Pandoc's convert from Markdown to typst string. 

Maybe you can tweak the template to use Block content like the default one, so that it works with escaped content. 

Handling this assumption of Markdown parsing and rendering to typst by replacing special character (using `replace()` function as shown in other reply) is probably a good way. 

I don't know what Quarto could do specifically. For you template, you could also imagine a Lua filter that would catch some metadata field parsed as Markdown, and mark them as `RawInline` for typst format after transforming as you want to be handled by your template. 

Example : making the email a rawInline instead of a parsed string 
````
> quarto pandoc -t native
test.test@test.me                 
^Z
[ Para [ Str "test.test@test.me" ] ]
> quarto pandoc -t native
`test.test@test.me`{=typst}
^Z
[ Para [ RawInline (Format "typst") "test.test@test.me" ] ]
````

Probably too complicated... 

Thanks for asking, it was nice to look deep into this to understand completely. We'll be able to answer more precisely if question comes again. 

Hope it helps

**Reply by charging1948** [reply]
It helps very much, thank you! Also thanks that you provided your insights into the matter!
